src/UniqueQueue.ts

```diff
@@ -1,71 +1,69 @@
-      !(arr.findIndex((arrEntry) => compFunc(entry, arrEntry)) === -1);
+    if (old?.completed) return;
 
-    this.queued = initialValues
-      ? initialValues.filter(this.duplicateFilterFunc)
-      : [];
-    this.completed = [];
-  }
+    if (!old || this.prefer?.(entry, old?.value)) {
+      this.uniqueMap.set(id, { value: entry, completed: false });
+    }
+    if (!old) this.queueArray.push(id);
+  };
 
-  // add new entries uniquely to the queued list, of they are not completed already
-  add = (...newEntries: T[]) =>
-    newEntries.filter(this.duplicateFilterFunc).forEach((entry) => {
-      // if the entry is not included in the list of completed entries
-      // add it to the queue, and remove duplicates afterwards
-      // --> this way the filter function can influence which of the duplicates is removed
-      //     (entries can differ but correspond to the same external data point
-      //      i.e. for cats we might have one entry { id: ..., weight: ... } and one entry { id: ..., name: ... }
-      //      and we can choose to keep the one that has the more relevant data)
-      // --> this could be improved in the future to also allow for merging of entries
-      if (!this.arrayIncludesEntry(this.completed, entry)) {
-        this.queued.unshift(entry);
-        this.queued = this.queued.filter(this.duplicateFilterFunc);
-      }
-    });
+  private getId: (arg: T) => string;
+  private prefer: ((a: T, b: T) => boolean) | undefined;
 
-  addCompleted = (...newEntries: T[]) => {
-    // remove them from being queued if they are flagged as completed
-    this.queued = this.queued.filter(
-      (entry) => !this.arrayIncludesEntry(newEntries, entry),
-    );
+  constructor(
+    getId: (arg: T) => string,
+    preferenceFunction?: (a: T, b: T) => boolean,
+    initialValues?: T[],
+  ) {
+    this.uniqueMap = new Map<
+      string,
+      { value: T; completed: false } | { completed: true }
+    >();
+    this.getId = getId;
+    this.prefer = preferenceFunction;
+    this.queueArray = [];
+    initialValues?.forEach(this.addSingleEntry);
+    this.completed = 0;
+  }
 
-    // add them uniquely to the completed entries
-    newEntries.filter(this.duplicateFilterFunc).forEach((entry) => {
-      if (!this.arrayIncludesEntry(this.completed, entry)) {
-        this.completed.push(entry);
-      }
-    });
+  // add new entries uniquely to the queued list, if they are not completed already
+  add = (...newEntries: T[]) => {
+    newEntries.forEach((entry) => this.addSingleEntry(entry));
   };
 
   getLength = () => ({
-    queued: this.queued.length,
-    completed: this.completed.length,
+    queued: this.queueArray.length,
+    completed: this.completed,
   });
 
-  next = () => {
-    const entry = this.queued.pop();
-    if (entry) this.completed.push(entry);
-    return entry;
+  next = (): T | undefined => {
+    const id = this.queueArray.pop();
+
+    console.log(this.getLength(), id);
```

Comment:
keeping intentionally?

Reply:
for now yes.

this allows us to see progress some information progress of the import in the console.

later on we might decide to log things differently (e.g. to logfiles as well), but for now this seemed reasonable to me

Reply:
very good